node/src/events.rs

```diff
@@ -0,0 +1,256 @@
+//! Events generated by `Node`
+
+use std::fmt;
+use std::panic::Location;
+use std::time::Duration;
+
+use instant::SystemTime;
+use serde::Serialize;
+use tokio::sync::broadcast;
+
+/// An error returned from the `EventReceiver::recv`.
+#[derive(Debug, thiserror::Error)]
+pub enum RecvError {
+    /// Node and all its event senders are closed.
+    #[error("Event channel closed")]
+    Closed,
+}
+
+/// An error returned from the `EventReceiver::try_recv`.
+#[derive(Debug, thiserror::Error)]
+pub enum TryRecvError {
+    /// The event channel is currently empty.
+    #[error("Event channel empty")]
+    Empty,
+    /// Node and all its event senders are closed.
+    #[error("Event channel closed")]
+    Closed,
+}
+
+/// A channel which users can subscribe for events.
+#[derive(Debug)]
+pub struct EventChannel {
+    tx: broadcast::Sender<NodeEventInfo>,
+}
+
+/// `EventPublisher` is used to broadcast events generated by [`Node`] to [`EventSubscriber`]s.
+///
+/// [`Node`]: crate::node::Node
+#[derive(Debug, Clone)]
+pub struct EventPublisher {
+    tx: broadcast::Sender<NodeEventInfo>,
+}
+
+/// `EventSubscriber` can be used by users to receive events from [`Node`].
+///
+/// [`Node`]: crate::node::Node
+#[derive(Debug)]
+pub struct EventSubscriber {
+    rx: broadcast::Receiver<NodeEventInfo>,
+}
+
+impl EventChannel {
+    /// Create a new `EventSubscription`.
+    pub fn new() -> EventChannel {
+        let (tx, _) = broadcast::channel(32);
```

Comment:
I think 32 can quickly become too little, but no idea what would be the best value there, 1024 maybe? :sweat_smile:

Reply:
changed to 1024